src/util/allocate_objects.cpp

```diff
@@ -0,0 +1,200 @@
+
+#include "allocate_objects.h"
+
+#include <util/c_types.h>
+#include <util/fresh_symbol.h>
+#include <util/pointer_offset_size.h>
+
+/// Installs a new symbol in the symbol table, pushing the corresponding
+/// `symbolt` object to the field `symbols_created` and emits to `assignments`
+/// a new assignment of the form `<target_expr> := address-of(new_object)`. The
+/// `allocate_type` may differ from `target_expr.type()`, e.g. for target_expr
+/// having type `int*` and `allocate_type` being an `int[10]`.
+///
+/// \param assignments: The code block to add code to.
+/// \param target_expr: The expression which we are allocating a symbol for.
+/// \param allocate_type: Type of the object allocated
+/// \param alloc_type: Allocation type (global, local or dynamic)
+/// \param symbols_created: Created symbols to be declared by the caller
+/// \return An expression denoting the address of the newly allocated object.
+exprt allocate_objectst::allocate_object(
+  code_blockt &assignments,
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const allocation_typet alloc_type,
+  std::vector<const symbolt *> &symbols_created)
+{
+  switch(alloc_type)
+  {
+  case allocation_typet::LOCAL:
+    return allocate_non_dynamic_object(
+      assignments, target_expr, allocate_type, false, symbols_created);
+    break;
+
+  case allocation_typet::GLOBAL:
+    return allocate_non_dynamic_object(
+      assignments, target_expr, allocate_type, true, symbols_created);
+    break;
+
+  case allocation_typet::DYNAMIC:
+    return allocate_dynamic_object(
+      assignments, target_expr, allocate_type, symbols_created);
+    break;
+  }
+
+  UNREACHABLE;
+}
+
+/// Generates code for allocating a non-dynamic local or global object.
+///
+/// \param assignments: Code block to which the necessary code is added
+/// \param target_expr: Expression to which the necessary memory will be
+///   allocated, its type should be pointer to `allocate_type`
+/// \param allocate_type: Type of the object allocated
+/// \param static_lifetime: If `true` a global object will be created, if
+///   `false` a local object will be created
+/// \param symbols_created: Created symbols to be declared by the caller
+/// \return An address_of_exprt of the newly allocated object.
+exprt allocate_objectst::allocate_non_dynamic_object(
+  code_blockt &assignments,
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const bool static_lifetime,
+  std::vector<const symbolt *> &symbols_created)
+{
+  symbolt &aux_symbol = get_fresh_aux_symbol(
+    allocate_type,
+    id2string(function),
+    "tmp_object_factory",
+    source_location,
+    symbol_mode,
+    symbol_table);
+
+  aux_symbol.is_static_lifetime = static_lifetime;
+  symbols_created.push_back(&aux_symbol);
+
+  exprt aoe = address_of_exprt(aux_symbol.symbol_expr());
+
+  if(!base_type_eq(allocate_type, target_expr.type().subtype(), ns))
+  {
+    aoe = typecast_exprt(aoe, target_expr.type());
+  }
```

Comment:
Not exactly the same (because it uses plain equality), but still: I'd go for `typecast_exprt::conditional_cast`.

Reply:
I also think it's likely that nearly-matching types (except at the very top level) may trip later passes into introducing (e.g.) unnecessary byte-extract operations

Reply:
Also this (potentially over-eager) typecast omission is a change in behaviour compared to the old code AFAICT; I'd suggest in that case it would want testing with something which is `base_type_eq` but not `type_eq` (e.g. structs whose fields differ in using a symbol vs. a struct type for one of the members)

Reply:
Ok, I've changed it to use `typecast_exprt::conditional_cast` (the original code used `ns.follow` on both types followed by comparing them via `==` to figure out whether a typecast should be introduced).